### ISSUE TITLE
AutoTrackRaymarine_pi  and  ocpn_plugin.h api 117

### ISSUE THREAD

**Issue by rgleason** [jongough/testplugin_pi#222]
Jon, 
I think the ocpn_plugin,h should be updated to the most current version.
For example, I am trying to build AutoTrackRaymarine  and it requires api 117.

Also, do you have any suggestions for me re this issue
https://github.com/rgleason/AutoTrackRaymarine_pi

**Comment by rgleason** [comment]
Ok this is the only plugin that needs it so far.  Do you have any idea what the error is?

**Comment by rgleason** [comment]
BTW Opencpn uses this ocpn_plugin.h

**Comment by jongough** [comment]
The intent of the file is to provide a consistent description of all interfaces to OCPN including definitions used for values and parameters. If OCPN wasn't using the latest version there would be an issue. However, my plugins used to use v1.13 and worked OK, then Plugin Manager changed things and destroyed backwards compatibility (lines 63-66 say it should be backwards compatible to avoid invalidating all plugins) and therefore required that I upgrade to 1.16. All plugins are now required to use 1.16 if they are to be able to interact with Plugin Manager. There should be no need to upgrade to the newer version 1.17 unless there is a need to access newer capabilities provided by it.

**Comment by jongough** [comment]
The plugin has names that don't match, casewise, for includes and calls. You will need to go through and change all uses to be either "AutoTrack*" or "Autotrack*". At least one of the files generated by wxFormbuilder has the wrong case or is case inconsistent. This may be OK on windows without case sensitivity, but most Linux platforms will fail if the case is not correct. "A" is not the same as "a".

**Comment by rgleason** [comment]
Thanks Jon, I will check that out. I think it was developed on windows.  The reason the ocpn_plugin.h is not fully backward compatible is the discontinuity of wxWidgets. Every time there is an update to wxWidgets, all backwards compatibility is lost.

**Comment by rgleason** [comment]
Jon, oddly, I have used notebook++ with "search in files" to find the differences in the case for AutoTrackRaymarine and I don't find that problem at all.

**Comment by jongough** [comment]
wxWidgets incompatibility is one thing which is easily handled at compile/build time by adjusting the version used. Most changes have no impact on the plugin. Changes to the OCPN API is what kills it. API's should not be modified in a way that makes them backwards incompatible. If this type of change is needed then a new API function should be introduced as this will not break existing code.

**Comment by jongough** [comment]
I have just tried to build a clean install from git and got this in the cmake part:
```
CMake Error at cmake/PluginSetup.cmake:27 (add_library):
  Cannot find source file:

    src/AutoTrackInfoUI.cpp

  Tried extensions .c .C .c++ .cc .cpp .cxx .cu .m .M .mm .h .hh .h++ .hm
  .hpp .hxx .in .txx
Call Stack (most recent call first):
  CMakeLists.txt:114 (include)
```
You can see it is looking for AutoTrackInfoUI.cpp but the module is AutotrackInfoUI.cpp, notice the difference case. This matters. There are at least 4 modules like this. Also, the "include" statements have similar case issues. This may build on Windows but will not build on any system that is case sensitive.

**Comment by rgleason** [comment]
OK so one of the problems is AutoTrackInfoUI  . Thanks, and I need to check the include statements.
I discovered that it would complete a build in windows when I use the original version of ocpn_plugin.h but not the updated v117. I don't know why that would happen.

**Comment by rgleason** [comment]
`wxWidgets incompatibility is one thing which is easily handled at compile/build time by adjusting the version used. Most changes have no impact on the plugin. Changes to the OCPN API is what kills it. API's should not be modified in a way that makes them backwards incompatible. If this type of change is needed then a new API function should be introduced as this will not break existing code.`

Jon I understand what you are saying, but wxWidgets itself is discontinuous requiring changes to code, I believe, which causes issues trying to maintain backwards compatibility.  All I know is that Pavel and bdbcat both advised that when we change wxWidgets version, we loose the capability of backwards compatibility.

**Comment by rgleason** [comment]
These instances all look ok to me.  I wonder what I am missing here?  - I guess I had best include cmakelists.txt etc too!
```
Search "autotrackinfoui" (4 hits in 4 files of 127 searched)
  C:\Users\fcgle\source\AutoTrackRaymarine_pi\src\AutotrackInfoUI.cpp (1 hit)
	Line 8: #include "AutotrackInfoUI.h"
  C:\Users\fcgle\source\AutoTrackRaymarine_pi\src\AutoTrackRaymarine_pi.cpp (1 hit)
	Line 38: #include "AutotrackInfoUI.h"
  C:\Users\fcgle\source\AutoTrackRaymarine_pi\src\Info.cpp (1 hit)
	Line 30: #include "AutotrackInfoUI.h"
  C:\Users\fcgle\source\AutoTrackRaymarine_pi\src\Info.h (1 hit)
	Line 33: #include "AutoTrackInfoUI.h"

Adding these

  C:\Users\fcgle\source\AutoTrackRaymarine_pi\CMakeLists.txt (2 hits)
	Line 149:       src/AutoTrackInfoUI.cpp
	Line 161:       src/AutoTrackInfoUI.h
  C:\Users\fcgle\source\AutoTrackRaymarine_pi\CMakeLists.txt.bak (2 hits)
	Line 149:       src/AutoTrackInfoUI.cpp
	Line 150:       src/AutoTrackInfoUI.h
  C:\Users\fcgle\source\AutoTrackRaymarine_pi\po\POTFILES.in (2 hits)
	Line 3: src/AutoTrackInfoUI.cpp
	Line 12: src/AutoTrackInfoUI.h
  C:\Users\fcgle\source\AutoTrackRaymarine_pi\po\POTFILES.in.test (2 hits)
	Line 3: src/AutoTrackInfoUI.cpp
	Line 12: src/AutoTrackInfoUI.h
  C:\Users\fcgle\source\AutoTrackRaymarine_pi\src\AutotrackInfoUI.cpp (1 hit)
	Line 8: #include "AutotrackInfoUI.h"
  C:\Users\fcgle\source\AutoTrackRaymarine_pi\src\AutoTrackRaymarine_pi.cpp (1 hit)
	Line 38: #include "AutotrackInfoUI.h"
  C:\Users\fcgle\source\AutoTrackRaymarine_pi\src\Info.cpp (1 hit)
	Line 30: #include "AutotrackInfoUI.h"
  C:\Users\fcgle\source\AutoTrackRaymarine_pi\src\Info.h (1 hit)
	Line 33: #include "AutoTrackInfoUI.h"

```

My eyesight is terrible when it comes to i or I  but this all looks ok to me.
Are there any others that you might recall?

**Comment by rgleason** [comment]
Well I checked all instances and they seemed ok as AutoTrackInfoUI.cpp, that failed so I changed the file name to have lower case "i" and that failed.  I've changed the "i" back to "I".  Now this is the error. So why and where should I look for the fix? 
```
CMake Error at cmake/PluginSetup.cmake:27 (add_library):
  Cannot find source file:
    src/AutoTrackInfoUI.cpp
```

**Comment by rgleason** [comment]
I got through the capitalization errors now I get this, HANDLE not been declared...

``
/run/build/AutoTrackRaymarine/src/NGT1Read.h:43:16: error: ‘HANDLE’ has not been declared
   int ReadNGT1(HANDLE handle);
                ^~~~~~
In file included from /run/build/AutoTrackRaymarine/src/PreferencesDialog.cpp:29:
/run/build/AutoTrackRaymarine/src/SerialPort.h:47:3: error: ‘HANDLE’ does not name a type; did you mean ‘wxHANDLER’?
   HANDLE m_hSerialin; //COM port handler
   ^~~~~~
   wxHANDLER
/run/build/AutoTrackRaymarine/src/SerialPort.h:49:44: error: ‘HANDLE’ has not been declared
   bool OpenSerialPort(wchar_t* pcCommPort, HANDLE* handle);
                                            ^~~~~~
/run/build/AutoTrackRaymarine/src/SerialPort.h:50:21: error: ‘HANDLE’ has not been declared
   void writeMessage(HANDLE handle, unsigned char command, const unsigned char * cmd, const size_t len);
                     ^~~~~~

```

**Comment by jongough** [comment]
You will need to go back to the last working (successful compile) version of the plugin and see what the 'HANDLE' is defined as. I tried on my machine, but I must have the wrong software installed for the 2019 version. I may try later if I have time.

**Comment by rgleason** [comment]
Jon, good point.  It builds for windows but I have found no way to "test" if it is working. VDR does not do it.

I will go back to the original and look.

**Comment by jongough** [comment]
wxWidgets is a breaking change, but that is normally handled by the build process with no changes needed in the code (there are instances where code changes are required, but most plugins should not need to do this). However, modifying an API into OCPN affects all plugins, i.e. the change to where plugin data is stored and the API changes around this. This break should not have occurred as a new API should have been used, or, if absolutely needed extra parameters (with defaults) could have been used. Changing anything in ocpn_plugin.h except at the end should not be acceptable. This means that most plugins should continue to work with a recompile/link.

**Comment by rgleason** [comment]
When I compared api it looked like the position or order of things changed and thers was an effort to clean up the order. I have no idea if that would affect anything.

I went back to Douwe's code, it is not complete, it is missing cmake files. It is also using sqlite and building "serial' both processes I dont know about  He also has a test area usibg python but I havebt figured out how to use that.
